### PR TITLE
Add ImportExternalTexture Video Test

### PR DESCRIPTION
**Pull request by bjjones**
<hr>

**Author checklist for test code/plans:**

- [x] All outstanding work is tracked with "TODO" in a test/file description or `.unimplemented()` on a test.
- [x] New helpers, if any, are documented using `/** doc comments */` and can be found via `helper_index.txt`.
- [x] (Optional, sometimes not possible.) Tests pass (or partially pass without unexpected issues) in an implementation. (Add any extra details above.)

**[Reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md) for test code/plans:** (Note: feel free to pull in other reviewers at any time for any reason.)

- [ ] The test path is reasonable, the [description](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) "describes the test, succinctly, but in enough detail that a reader can read only the test plans in a file or directory and evaluate the completeness of the test coverage."
- [ ] Tests appear to cover this area completely, except for outstanding TODOs. Validation tests use control cases.
    (This is critical for coverage. Assume anything without a TODO will be forgotten about forever.)
- [ ] Existing (or new) test helpers are used where they would reduce complexity.
- [ ] TypeScript code is readable and understandable (is unobtrusive, has reasonable type-safety/verbosity/dynamicity).

### CODE REVIEW

**Comment by bjjones** [comment]
I'm a little confused on what test goes where for this. This file existed already and seemed to fit - but I've still got a couple more tests that seem to fit better elsewhere like:
- Ensure that external texture destruction occurs via microtask
- Ensure that using a destroyed external texture results in error
- Ensure that using an invalid HTMLVideoElement results in error

I feel like these^ tests belong in the normal /validation/importExternalTexture.spec.ts and /operation/importExternalTexture.spec.ts files. Or do you think I should just stick everything in this file?

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (kainino0x) [inline src/webgpu/web_platform/external_texture/video.spec.ts:50]: Is it necessary to call this twice? One `await video.play()` seems enough, but if it's not, leave a comment about it.

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (kainino0x) [inline src/webgpu/web_platform/external_texture/video.spec.ts:40]: I don't think this URL is going to work when we're in WPT. Fortunately, for web_platform, tests we don't have to worry about other runtimes (cmdline (node) and deno runtimes), but we still have both the standalone and wpt runtimes. And the wpt runtime is the one we use on automated testing, so we'll want it to work.

We may need some kind of base URL (pointing to the root of the src/ or out/ or out-wpt/ directory or whatever), passed in from the runtime, so that we know how to grab these resources. Happy to chat more about this, or look into it (but may take me some time to get to it).

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (kainino0x) [inline src/webgpu/web_platform/external_texture/video.spec.ts:49]: The state of the art in getting videos to play for tests can usually be found in `startPlayingAndWaitForVideo`:
https://github.com/KhronosGroup/WebGL/blob/main/sdk/tests/js/webgl-test-utils.js
So it might be worth adding a link here to that, in case we ever need to reference it again.

I'm not sure whether we should port in everything that's there right now. It may be the safest bet.

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (kainino0x) [inline src/webgpu/web_platform/external_texture/video.spec.ts:139]: I hope these are the right colors once color management is taken into account! I don't actually have any clue...

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (bjjones) [reply src/webgpu/web_platform/external_texture/video.spec.ts]: This was a mistake. Only one is needed.

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (bjjones) [reply src/webgpu/web_platform/external_texture/video.spec.ts]: At least 3 of these tests pass (not sure about mp4...haven't tried with an mp4-enabled build). I'm actually sampling a few pixels away from the edge, which allows me to sample exactly red/green. Added a comment for this.

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (bjjones) [reply src/webgpu/web_platform/external_texture/video.spec.ts]: I think it makes sense to add in at least a portion of the helper - it seems like it's going to require running each test as a callback, so may as well start in that format to avoid churn later.

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (bjjones) [reply src/webgpu/web_platform/external_texture/video.spec.ts]: I see - so it looks like I can acquire `window.location.href` within common/runtime/wpt.ts and then pass it into `run()` - right? Alternatively is there any reason I can't just use `window.location.href` directly from the test?

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (kainino0x) [reply src/webgpu/web_platform/external_texture/video.spec.ts:138]: I'm thinking about color profiles/ICC profiles. I don't really know how they work with videos. I assume these test videos are "untagged" insofar as a video can be, so I think these expectations are correct.

(The comment is good though, thanks)

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (kainino0x) [reply src/webgpu/web_platform/external_texture/video.spec.ts:40]: Unfortunately window.location.href by itself won't let us figure out the location of the resources directory. The file paths look like this:

- In standalone: `/standalone/` -> `/out/resources/`
- In the actual WPT build: `/webgpu/cts.html` -> `/webgpu/resources/`
- In Chromium: `/wpt_internal/webgpu/cts.html` -> `/gen/third_party/webgpu-cts/src/`

In order to deal with this for the rest of the generated files, we point to the appropriate build directory from the template:
- https://github.com/gpuweb/cts/blob/main/src/common/templates/cts.html#L32
- https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/webgpu/ctshtml-template.txt;l=24;drc=6b69efe99ef855e2ac953b37a31063d07ecfee9a

So I think we are going to have to add something like `loadWebGPUExpectations` to point to the root of the build directory.

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (kainino0x) [reply src/webgpu/web_platform/external_texture/video.spec.ts]: I think that makes sense. It could be a promise instead of a callback - but I suppose when requestVideoFrameCallback is supported, the callback is probably more likely to give us exactly the right frame (rather than a later one).

Comment (bjjones) [reply src/webgpu/web_platform/external_texture/video.spec.ts]: I've been trying to figure out how use the their function directly - but I'm not very experienced with JS and having issues with async execution. I've updated the PR with startPlayingAndWaitForVideo, however it doesn't function correctly. I find that finalize() is being called before the test callback occurs, which is obviously wrong. Maybe take a look and see if you have a suggestion? If I had a way for startPlayingAndWaitForVideo to wait until maybeCallCallback has finished, we could use this.

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (bjjones) [reply src/webgpu/web_platform/external_texture/video.spec.ts]: Resolved.

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (bjjones) [reply src/webgpu/web_platform/external_texture/video.spec.ts]: Resolved.

**Comment by bjjones** [comment]
Comments resolved. Thanks for the help.

**Comment by shaoboyan** [comment]
@bjjones would it possible to cover the usage of external texture in compute shader too ?

**Comment by bjjones** [comment]
> @bjjones would it possible to cover the usage of external texture in compute shader too ?

Added.

**Comment by bjjones** [comment]
@kainino0x I added a check for microtask destruction in the sampling test. Lmk if you'd rather see it split into it's own test (it may also be more appropriate as a validation test?)

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (kainino0x) [inline src/webgpu/web_platform/external_texture/video.spec.ts:38]: Incomplete?

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (kainino0x) [inline src/webgpu/web_platform/external_texture/video.spec.ts:144]: nit: I think if there are corner artifacts it would be due to compression, not color management
```suggestion
      // Bottom right corner should be green. Sample a few pixels away from the edges to avoid
      // compression artifacts.
```

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (kainino0x) [inline src/webgpu/web_platform/external_texture/video.spec.ts:214]: Ideally copy the data into a storage texture and check the contents? That way we can be sure the sampling is working in compute shaders when we implement the 0-copy impl (which might differ slightly for compute shaders since iirc texture sampling of the underlying textures won't be possible in compute).

If not implementing this now, please leave a TODO before landing the PR.

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (kainino0x) [inline src/webgpu/web_platform/external_texture/video.spec.ts:240]: This number is too big for uint32array, is this meant to be:
```suggestion
      const resultData = new Uint32Array([80, 128]);
```
?

**Review by kainino0x — APPROVED** [review]
> @kainino0x I added a check for microtask destruction in the sampling test. Lmk if you'd rather see it split into it's own test (it may also be more appropriate as a validation test?)

Would probably be ideal to have it be a separate test if it's not too hard. Then browsers can skip it if their implementation or the test turns out to be flaky. (The timing of this is potentially hard to test reliably.)
I'm neutral on validation vs operation test. It fits equally well into both, I think.

**Comment by kainino0x** [comment]
LGTM, but I'd like to review the diff if you split the destruction test or enhance the compute test. Or feel free to land this with TODOs and do that in a new PR.

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (bjjones) [reply src/webgpu/web_platform/external_texture/video.spec.ts]: Done.

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (bjjones) [reply src/webgpu/web_platform/external_texture/video.spec.ts:215]: Added TODO

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (bjjones) [reply src/webgpu/web_platform/external_texture/video.spec.ts]: Done.

File: src/webgpu/web_platform/external_texture/video.spec.ts
Comment (bjjones) [reply src/webgpu/web_platform/external_texture/video.spec.ts]: Removed.

**Comment by bjjones** [comment]
Will merge after this passes. Will do a follow  up PR to address TODOs.

**Pull request merged**
